pkg/tar/tar.go

```diff
@@ -38,10 +38,10 @@ var ErrNotSupportedPlatform = errors.New("platform and architecture is not suppo
 // root of the tar file and should be cleaned (for example using filepath.Clean)
 type PathWhitelistMap map[string]struct{}
 
-// extractTar extracts a tarball (from a tar.Reader) into the root directory
-// if pwl is not nil, only the paths in the map are extracted.
-// If overwrite is true, existing files will be overwritten.
-func extractTar(tr *tar.Reader, overwrite bool, pwl PathWhitelistMap, uidRange *uid.UidRange) error {
+// ExtractTarInsecure extracts a tarball (from a tar.Reader) into the root
+// directory if pwl is not nil, only the paths in the map are extracted. If
+// overwrite is true, existing files will be overwritten.
+func ExtractTarInsecure(tr *tar.Reader, overwrite bool, pwl PathWhitelistMap, uidRange *uid.UidRange) error {
```

Comment:
`uidRange *uid.UidRange` 
changes to
`cb FilePermissionsEditor`

where we have a generic interface that extractors can fulfill
`type FilePermissionsEditor func(path string, fi os.FileInfo) error`
(stupid name, hopefully you can come up with a better one)

line 169 becomes
```
if err := cb(p, fi); err != nil { 
   return err 
}
```

what is currently in line 169-184 becomes 
```
func makeUidShiftingFilePermissionsEditor(uidRange *uid.UidRange) FilePermissionsEditor
... 
}
```
then you define a 
```
func MaybeExtractThingiesAsTheEncodedUserIfWereRootOtherwiseJustExtractAsOurselves(path string, fi os.FileInfo) error {
    ....
}
```